ext/mbstring/mbstring.c

```diff
@@ -2941,6 +2941,153 @@ PHP_FUNCTION(mb_strtolower)
+{
+	for (size_t i = 0; i < characters_len; i++) {
+		if (w == characters[i]) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static zend_string* trim_each_wchar(zend_string *str, zend_string *what, int mode, const mbfl_encoding *enc)
+{
+	unsigned char *in = (unsigned char*)ZSTR_VAL(str);
+	unsigned char *what_in = (unsigned char*)ZSTR_VAL(what);
+	uint32_t wchar_buf[128];
+	uint32_t what_wchar_buf[128];
+	size_t in_len = ZSTR_LEN(str);
+	size_t out_len = 0;
+	size_t what_out_len = 0;
+	size_t what_len = ZSTR_LEN(what);
+	unsigned int state = 0;
+	long left = 0;
+	long right = 0;
+	size_t total_len = 0;
+
+	if (what_len == 0) {
+		return str;
+	}
+
+	while (in_len) {
+		out_len = enc->to_wchar(&in, &in_len, wchar_buf, 128, &state);
+		ZEND_ASSERT(out_len <= 128);
+		total_len += out_len;
+
+		while (what_len) {
+			what_out_len = enc->to_wchar(&what_in, &what_len, what_wchar_buf, 128, &state);
+			ZEND_ASSERT(what_out_len <= 128);
+
+			if (mode & 1) {
+				for (size_t i = 0; i < out_len; i++) {
+					uint32_t w = wchar_buf[i];
+					if (is_trim_wchar(w, what_wchar_buf, what_out_len)) {
+						left += 1;
+					} else {
+						what_in -= ZSTR_LEN(what);
+						what_len = ZSTR_LEN(what);
+						goto mb_ltrim_end;
+					}
+				}
+			}
+		}
+
+		what_in -= ZSTR_LEN(what);
+		what_len = ZSTR_LEN(what);
+	}
+
+mb_ltrim_end:
+
+	in = (unsigned char*)ZSTR_VAL(str);
+	in_len = ZSTR_LEN(str);
+
+	while (in_len) {
+		out_len = enc->to_wchar(&in, &in_len, wchar_buf, 128, &state);
+		ZEND_ASSERT(out_len <= 128);
+
+		while (what_len) {
+			what_out_len = enc->to_wchar(&what_in, &what_len, what_wchar_buf, 128, &state);
+			ZEND_ASSERT(what_out_len <= 128);
+
+			if (mode & 2) {
+				for (size_t j = 0; j < out_len; j++) {
+					uint32_t w = wchar_buf[j];
+					if (is_trim_wchar(w, what_wchar_buf, what_out_len)) {
+						right -= 1;
+					} else {
+						right = 0;
+					}
+				}
+			}
+		}
+
+		what_in -= ZSTR_LEN(what);
+		what_len = ZSTR_LEN(what);
+	}
+
+	zend_string_release(what);
+
+	return mb_get_substr(str, left, total_len + (right - left), enc);
+}
+
+static zend_string* mb_trim_default_chars(zend_string *str, int mode, const mbfl_encoding *enc)
+{
+	uint32_t trim_default_chars[] = {
+		0x20, 0x0C, 0x0A, 0x0D, 0x09, 0x0B, 0x00, 0xA0, 0x1680,
+		0x2000, 0x2001, 0x2002, 0x2003, 0x2004, 0x2005, 0x2006, 0x2007,
+		0x2008, 0x2009, 0x200A, 0x2028, 0x2029, 0x202F, 0x205F, 0x3000,
+		0x85, 0x180E
+	};
+	size_t trim_default_chars_length = sizeof(trim_default_chars) / sizeof(uint32_t);
+	mb_convert_buf buf;
+	mb_convert_buf_init(&buf, trim_default_chars_length, MBSTRG(current_filter_illegal_substchar), MBSTRG(current_filter_illegal_mode));
+
+	enc->from_wchar(trim_default_chars, trim_default_chars_length, &buf, true);
+
+	return trim_each_wchar(str, mb_convert_buf_result(&buf, enc), mode, enc);
+}
+
+static zend_always_inline void php_do_mb_trim(INTERNAL_FUNCTION_PARAMETERS, int mode)
```

Comment:
Can this mode parameter made an enum, as just seeing a magic number is kinda confusing.

Reply:
Sure, I followed trim function, that `php_do_trim` of `mode` is commented below.

```c
/* {{{ php_trim_int()
 * mode 1 : trim left
 * mode 2 : trim right
 * mode 3 : trim left and right
 * what indicates which chars are to be trimmed. NULL->default (' \t\n\r\v\0')
 */
```

However, I think make sense make an enum. Because these code is since 2001.

Reply:
Added f4b3c5b879d7550445dd37e4311be69f0230b96f.